monad-stack/src/main/java/com/github/jorgecastillo/kotlinandroid/data/datasource/remote/MarvelNetworkDataSource.kt

```diff
@@ -25,24 +25,26 @@ import kotlinx.coroutines.experimental.async
  * required execution context.
  */
 
-fun fetchAllHeroes() = Reader.ask<IdHK, GetHeroesContext>().map({ ctx ->
+fun fetchAllHeroes() = Reader.ask<ForId, GetHeroesContext>().map({ ctx ->
   IO.monadSuspend().bindingCatch {
-    runInAsyncContext(
+    val result = runInAsyncContext(
         f = { queryForHeroes(ctx) },
         onError = { IO.raiseError<List<CharacterDto>>(it) },
         onSuccess = { IO.pure(it) },
         AC = ctx.threading
     ).bind()
+    result.bind()
   }
 })
 
-fun fetchHeroDetails(heroId: String) = Reader.ask<IdHK, GetHeroDetailsContext>().map({ ctx ->
+fun fetchHeroDetails(heroId: String) = Reader.ask<ForId, GetHeroDetailsContext>().map({ ctx ->
   IO.monadSuspend().bindingCatch {
-    runInAsyncContext(
+    val result = runInAsyncContext(
         f = { queryForHero(ctx, heroId) },
         onError = { IO.raiseError<List<CharacterDto>>(it) },
         onSuccess = { IO.pure(it) },
         AC = ctx.threading).bind()
+    result.bind()
```

Comment:
same. I'm starting to think this has a rationale. Is the function bind() forcing the function to resolve and the result bind forcing the effects returned by the function to resolve? Are both different things?

Reply:
`val result : Kind<F, CharacterDto>` so in order to extract the `CharacterDto` you need to call `bind` again. Due to how `runInAsyncContext` you need to wrap it `IO` again to deal with the error `IO.raiseError<List<CharacterDto>>(it)`.

Can be rewritten to if I am not mistaken.
```kotlin
fun fetchHeroDetails(heroId: String) = Reader.ask<ForId, GetHeroDetailsContext>().map({ ctx ->
    IO.monadSuspend().bindingCatch {
        ctx.threading.async<List<CharacterDto>> {
            Try { queryForHero(ctx, heroId) }
                    .fold(
                            { it.left() },
                            { it.right() }
                    )
        }.bind()
    }
})
```

And that could even be reduced to since it's in a `bindingCatch` block.
```kotlin
fun fetchHeroDetails(heroId: String) = Reader.ask<ForId, GetHeroDetailsContext>().map({ ctx ->
    IO.monadSuspend().bindingCatch {
        ctx.threading.async<List<CharacterDto>> {
            queryForHero(ctx, heroId)
        }.bind()
    }
})
```

(Correct me if I am wrong please.)

Reply:
I think you lack there the coroutine running it on the common pool to avoid the Android `NetworkOnMainThreadException`. That's the main purpose of the `runrunInAsyncContext()` function that runs what you have there but inside a forced coroutine dispatcher. (Arrow still not provides support for coroutine dispatchers, even though it's ongoing by @pakoito)